### ISSUE TITLE
[idea - needs work][discussion] `z.satisfies` companion to `z.infer`

### ISSUE THREAD

**Issue by WillsterJohnson** [colinhacks/zod#2473]
TLDR;

Allows for overriding the output of `z.infer` while still being constrained to it's requirements.

```ts
type satisfies<T extends z.infer<Zod>, Zod extends z.ZodType> = T extends z.infer<Zod> ? T : never;
```

## Why

Due to a combination of the way Zod is built and the limitations of TS's type system, types produced by `z.infer` can be difficult to read. While intellisense does a lot of the legwork for us, there are some cases where it's worth the extra work of manually writing a type just to clean up TS's inference. This becomes increasingly true for larger unions and non-trivial object structures.

## How

```ts
type satisfies<T extends z.infer<Zod>, Zod extends z.ZodType> = T extends z.infer<Zod> ? T : never;
```

The generic `Zod` here must be a valid input to `z.infer`, and the generic `T` must satisfy that inferred constraint. This leads to some interesting behavior beyond simply working as a `satisifes` clause for the user-defined type. Some of these are desirable, others less so, and ultimately this does *not* enforce everything that the schema enforces.
Therefore, the purpose of this is less "it guides you as you write your own inferred type, blocking anything which is not permitted" and more of a "you know what you're doing and recognise this is **satisifies**, not **implements**".

An implements type could be created with a little bit more work, though I doubt that it would be used much. The kind of user who would need a strict constraint when manually defining the type of their schema is the kind of user who more than likely wouldn't want to go down that route to begin with.

## Key Benefit

The major benefit of this I see is that Zod schemas can be made infinitely generic. For example, some `z.any()` can instead be made generic, allowing library authors or (more likely) users to create their own type safety for situations where the internal workings don't care what data is at a certain place, but consumers want to ensure they adhere to some structure or other when manipulating it.

```ts
const mySchema = z.object({
  foo: z.string(),
  bar: z.string(),
  baz: z.record(z.string(), z.any())
});
type MySchemaTyped<T extends Record<string, any> = Record<string, any>> = z.satisfies<{
  foo: string;
  bar: string;
  baz: T;
}, typeof mySchema>

// consumer
const myData = mySchema.parse({
  foo: "foo",
  bar: "bar",
  baz: {
    thisIsNot: "assignable to `Thing`",
//  ^^^^^^^^^ ts(2322)
  },
} satisfies MySchemaTyped<Thing>) as MySchemaTyped<Thing>;
```

This could be more user-friendly and less verbose with a modified `parse` signature. It would no longer allow any input, however it would use TS to validate that the input matches the additional constraint. This would not effect the runtime at all; this is a self-imposed restriction rather than a validation restriction.

```ts
// additional parse signature example
    parse(data: unknown, params?: Partial<ParseParams>): Output;
    parse<EnforceInput extends Output = Output>(data: EnforceInput, params?: Partial<ParseParams>): T;

const myData = mySchema.parse<MySchemaTyped<Thing>>({
  foo: "foo",
  bar: "bar",
  baz: {
    thisIsNot: "assignable to `Thing`",
//  ^^^^^^^^^ ts(2322)
  },
});
```

**Comment by JacobWeisenburger** [comment]
Is this helpful for you?
https://github.com/colinhacks/zod/issues/1748
